src/nrnpython/nrnpy_p2h.cpp

```diff
@@ -919,19 +919,29 @@ static Object* py_alltoall_type(int size, int type) {
                 delete[] sdispl;
 
             if (rcnt[0]) {
-                nb::object po = unpickle(r);
-                pdest = po.release().ptr();
+                pdest = unpickle(r);
             } else {
-                pdest = Py_None;
-                Py_INCREF(pdest);
+                pdest = nb::none();
             }
 
             delete[] rcnt;
             assert(rdispl == NULL);
         }
     }
-    Object* ho = nrnpy_po2ho(pdest);
-    Py_XDECREF(pdest);
+
+    Object* ho = nrnpy_po2ho(pdest.ptr());
+
+    // The problem here is when `pdest` is a HOC object. In that case `ho` has `refcount == 2` but
+    // must be returned with refcount 0. The `ho` is contained in the `pdest` (which is why the HOC
+    // refcount is 2) and will be destroyed along with the `pdest` if its reference count is 0.
+    //
+    // Therefore, we must first DECREF `pdest` and then decrement `ho->refcount`. If we do
+    // it the other way around the `Py_DECREF` causes the Python object to be deallocated,
+    // as part of that deallocation method, the contained HOC object is dereferenced, and because
+    // it'll have a reference count of 0 it's also be deallocated.
```

Comment:
Michael, could you please read this comment and check if it makes any sense? This changed looked harmless without these lines. However, it would reliably segfault on my machine. When I started debugging this was the closest I got to an explanation.

Fernando pointed out that the first attempt that the description was flawed, so I updated it and move the question.